File: kubectl-cloudflow/verify/verifiedblueprint.go
package verify

import (
	"fmt"
	"strings"

	"github.com/lightbend/cloudflow/kubectl-cloudflow/cloudflowapplication"
)

// SchemaDescriptor is a domain expressive typedef
type SchemaDescriptor cloudflowapplication.InOutletSchema

// OutletDescriptor is a domain expressive typedef
type OutletDescriptor cloudflowapplication.InOutlet

// InletDescriptor is a domain expressive typedef
type InletDescriptor cloudflowapplication.InOutlet

type VerifiedStreamlet struct {
	name       string
	descriptor StreamletDescriptor
}

type VerifiedPortPath struct {
	streamletRef string
	portName     *string
}

type VerifiedPort struct {
	portName         string
	schemaDescriptor cloudflowapplication.InOutletSchema
}

type VerifiedInlet struct {
	VerifiedPort
	streamlet VerifiedStreamlet
}

type VerifiedOutlet struct {
	VerifiedPort
	streamlet VerifiedStreamlet
}

type VerifiedStreamletConnection struct {
	verifiedOutlet VerifiedOutlet
	verifiedInlet  VerifiedInlet
	label          *string
}

type VerifiedBlueprint struct {
	streamlets  []VerifiedStreamlet
	connections []VerifiedStreamletConnection
}

func (v *VerifiedStreamlet) outlet(outlet OutletDescriptor) VerifiedOutlet {
	return VerifiedOutlet{
		VerifiedPort{
			outlet.Name,
			outlet.Schema,
		},
		*v,
	}
}

func (v *VerifiedStreamlet) inlet(inlet InletDescriptor) VerifiedOutlet {
	return VerifiedOutlet{
		VerifiedPort{
			inlet.Name,
			inlet.Schema,
		},
		*v,
	}
}

// ToString string representation of a VerifiedPortPath
func (v VerifiedPortPath) ToString() string {
	if v.portName == nil {
		return v.streamletRef
	}
	return fmt.Sprintf("%s.%s", v.streamletRef, *v.portName)
}

// NewVerifiedPortPath constructs a VerifiedPortPath out of a string
func NewVerifiedPortPath(portPath string) (*VerifiedPortPath, *InvalidPortPath) {
	var trimmed = strings.TrimSpace(portPath)
	var splitF = func(c rune) bool {
		return c == '.'
	}
	var parts = strings.FieldsFunc(trimmed, splitF)

	if strings.HasPrefix(trimmed, ".") {
		return nil, &InvalidPortPath{
			PortPathError: PortPathError{},
			path:          portPath}
	} else if len(parts) == 1 && !strings.HasSuffix(portPath, ".") {
		return &VerifiedPortPath{parts[0], nil}, nil
	} else if len(parts) >= 2 {
		var portName = parts[len(parts)-1]
		var streamletNamePart = parts[:len(parts)-1]
		var streamletRef = mkString(streamletNamePart, ".")
		if len(streamletRef) == 0. {
			return nil, &InvalidPortPath{
				PortPathError: PortPathError{},
				path:          portPath}
		} else if len(portName) == 0 {
			return nil, &InvalidPortPath{
				PortPathError: PortPathError{},
				path:          portPath}
		} else {
			return &VerifiedPortPath{parts[0], &parts[1]}, nil
		}
	} else {
		return nil, &InvalidPortPath{
			PortPathError: PortPathError{},
			path:          portPath}
	}
}

func (v *VerifiedInlet) portPath() VerifiedPortPath {
	return VerifiedPortPath{
		v.streamlet.name,
		&v.VerifiedPort.portName,
	}
}

func (v *VerifiedOutlet) portPath() VerifiedPortPath {
	return VerifiedPortPath{
		v.streamlet.name,
		&v.VerifiedPort.portName,
	}
}

func (v *VerifiedOutlet) matches(outletDescriptor OutletDescriptor) bool {
	return outletDescriptor.Name == v.portName && outletDescriptor.Schema.Fingerprint == v.schemaDescriptor.Fingerprint
}

// FindVerifiedOutlet finds the VerifiedOutlet corresponding to the outletPortPath from the list of
// verifiedStreamlets
func FindVerifiedOutlet(verifiedStreamlets []VerifiedStreamlet, outletPortPath string) (*VerifiedOutlet, BlueprintProblem) {
	verifiedPortPath, err := NewVerifiedPortPath(outletPortPath)

	if err != nil {
		return nil, err
	}

	var foundVerifiedStreamlet *VerifiedStreamlet
	for _, verifiedStreamlet := range verifiedStreamlets {
		if verifiedStreamlet.name == verifiedPortPath.streamletRef {
			foundVerifiedStreamlet = &verifiedStreamlet
			break
		}
	}
	if foundVerifiedStreamlet == nil {
		return nil, PortPathNotFound{
			path: outletPortPath,
		}
	}
	if verifiedPortPath.portName == nil && len(foundVerifiedStreamlet.descriptor.Outlets) > 1 {
		var outletsToMap = foundVerifiedStreamlet.descriptor.Outlets
		var suggestions []VerifiedPortPath

		for _, outletToMap := range outletsToMap {
			finalNameStr := outletToMap.Name
			suggestions = append(suggestions, VerifiedPortPath{
				streamletRef: verifiedPortPath.streamletRef,
				portName:     &finalNameStr})
		}
		return nil, PortPathNotFound{
			path:        outletPortPath,
			suggestions: suggestions,
		}
	}
	var portPath *VerifiedPortPath
	if verifiedPortPath.portName == nil && len(foundVerifiedStreamlet.descriptor.Outlets) == 1 {
		finalNameStr := foundVerifiedStreamlet.descriptor.Outlets[0].Name
		portPath = &VerifiedPortPath{
			streamletRef: verifiedPortPath.streamletRef,
			portName:     &finalNameStr}

	} else {
		portPath = verifiedPortPath
	}

	var foundOutlet *VerifiedOutlet
	for _, outlet := range foundVerifiedStreamlet.descriptor.Outlets {
		if portPath.portName != nil && outlet.Name == *portPath.portName {
			foundOutlet = &VerifiedOutlet{
				streamlet: *foundVerifiedStreamlet,
				VerifiedPort: VerifiedPort{
					portName:         outlet.Name,
					schemaDescriptor: outlet.Schema,
				},
			}
			break
		}
	}

	if foundOutlet != nil {
		return foundOutlet, nil
	}
	return nil, PortPathNotFound{
		path: outletPortPath,
	}
}

// FindVerifiedInlet finds the VerifiedInlet corresponding to the inletPortPath from the list of
// verifiedStreamlets
func FindVerifiedInlet(verifiedStreamlets []VerifiedStreamlet, inletPortPath string) (*VerifiedInlet, BlueprintProblem) {
	verifiedPortPath, err := NewVerifiedPortPath(inletPortPath)

	if err != nil {
		return nil, err
	}
	var found *VerifiedStreamlet
	for _, verifiedStreamlet := range verifiedStreamlets {
		if verifiedStreamlet.name == verifiedPortPath.streamletRef {
			found = &verifiedStreamlet
			break
		}
	}
	if found == nil {
		return nil, PortPathNotFound{
			path: inletPortPath,
		}
	}
	if verifiedPortPath.portName == nil && len(found.descriptor.Inlets) > 1 {
		var inletsToMap = found.descriptor.Inlets
		var suggestions []VerifiedPortPath

		for _, inletToMap := range inletsToMap {
			finalNameStr := inletToMap.Name
			suggestions = append(suggestions, VerifiedPortPath{
				streamletRef: verifiedPortPath.streamletRef,
				portName:     &finalNameStr})
		}
		return nil, PortPathNotFound{
			path:        inletPortPath,
			suggestions: suggestions,
		}
	}
	var portPath *VerifiedPortPath
	if verifiedPortPath.portName == nil && len(found.descriptor.Inlets) == 1 {
		finalNameStr := found.descriptor.Inlets[0].Name
		portPath = &VerifiedPortPath{
			streamletRef: verifiedPortPath.streamletRef,
			portName:     &finalNameStr}

	} else {
		portPath = verifiedPortPath
	}

	var foundInlet *VerifiedInlet
	for _, inlet := range found.descriptor.Inlets {
		if portPath.portName != nil && inlet.Name == *portPath.portName {
			foundInlet = &VerifiedInlet{
				streamlet: *found,
				VerifiedPort: VerifiedPort{
					portName:         inlet.Name,
					schemaDescriptor: inlet.Schema,
				},
			}
			break
		}
	}

	if foundInlet != nil {
		return foundInlet, nil
	}
	return nil, PortPathNotFound{
		path: inletPortPath,
	}
}

File: core/cloudflow-blueprint/src/main/scala/cloudflow/blueprint/VerifiedBlueprint.scala
/*
 * Copyright (C) 2016-2020 Lightbend Inc. <https://www.lightbend.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cloudflow.blueprint

case class VerifiedBlueprint(
    streamlets: Vector[VerifiedStreamlet],
    connections: Vector[VerifiedStreamletConnection]
) {
  def findOutlet(outletPortPath: String): Either[PortPathError, VerifiedOutlet] =
    VerifiedOutlet.find(streamlets, outletPortPath)
}

object VerifiedPortPath {
  def apply(portPath: String): Either[PortPathError, VerifiedPortPath] = {
    val trimmed = portPath.trim()
    val parts   = trimmed.split("\\.").filterNot(_.isEmpty).toVector
    if (trimmed.startsWith(".")) {
      Left(InvalidPortPath(portPath))
    } else if (parts.size == 1 && !portPath.endsWith(".")) {
      Right(VerifiedPortPath(parts.head, None))
    } else if (parts.size >= 2) {
      val portName          = parts.last
      val streamletNamePart = parts.init
      val streamletRef      = streamletNamePart.mkString(".")
      if (streamletRef.isEmpty) Left(InvalidPortPath(portPath))
      else if (portName.isEmpty) Left(InvalidPortPath(portPath))
      else Right(VerifiedPortPath(streamletRef, Some(portName)))
    } else {
      Left(InvalidPortPath(portPath))
    }
  }
}

final case class VerifiedPortPath(streamletRef: String, portName: Option[String]) {
  override def toString = portName.fold(streamletRef)(port ⇒ s"$streamletRef.$port")
}

final case class VerifiedStreamlet(
    name: String,
    descriptor: StreamletDescriptor,
    inletRefs: Vector[InletRef] = Vector.empty[InletRef],
    outletRefs: Vector[OutletRef] = Vector.empty[OutletRef]
) {
  def outlet(outlet: OutletDescriptor) = VerifiedOutlet(this, outlet.name, outlet.schema)
  def inlet(inlet: InletDescriptor)    = VerifiedInlet(this, inlet.name, inlet.schema)
}

final case class VerifiedStreamletConnection(verifiedOutlet: VerifiedOutlet, verifiedInlet: VerifiedInlet, label: Option[String] = None)

sealed trait VerifiedPort {
  def portName: String
  def schemaDescriptor: SchemaDescriptor
}

object VerifiedOutlet {
  def find(
      verifiedStreamlets: Vector[VerifiedStreamlet],
      outletPortPath: String
  ): Either[PortPathError, VerifiedOutlet] =
    for {
      verifiedPortPath ← VerifiedPortPath(outletPortPath)
      verifiedOutlet ← verifiedStreamlets
        .find(_.name == verifiedPortPath.streamletRef)
        .toRight(PortPathNotFound(outletPortPath))
        .flatMap { verifiedStreamlet ⇒
          // a port path can leave out the outlet
          val portNameFound = verifiedPortPath.portName.orElse {
            if (verifiedStreamlet.descriptor.outlets.size == 1) Some(verifiedStreamlet.descriptor.outlets.head.name)
            else None
          }
          verifiedStreamlet.outletRefs
            .find(outletRef => Some(outletRef.outletName) == portNameFound)
            .map { outletRef =>
              Left(PortAlreadyBoundToTopic(outletPortPath, outletRef.topic))
            }
            .getOrElse {
              if (verifiedPortPath.portName.isEmpty && verifiedStreamlet.descriptor.outlets.size > 1) {
                val suggestions = verifiedStreamlet.descriptor.outlets.map(outlet ⇒ verifiedPortPath.copy(portName = Some(outlet.name)))
                Left(PortPathNotFound(outletPortPath, suggestions))
              } else {
                val portPath = if (verifiedPortPath.portName.isEmpty && verifiedStreamlet.descriptor.outlets.size == 1) {
                  verifiedPortPath.copy(portName = Some(verifiedStreamlet.descriptor.outlets.head.name))
                } else verifiedPortPath

                verifiedStreamlet.descriptor.outlets
                  .find(outlet ⇒ Some(outlet.name) == portPath.portName)
                  .map(outletDescriptor ⇒ VerifiedOutlet(verifiedStreamlet, outletDescriptor.name, outletDescriptor.schema))
                  .toRight(PortPathNotFound(outletPortPath))
              }
            }
        }
    } yield verifiedOutlet
}

object VerifiedInlet {
  def find(
      verifiedStreamlets: Vector[VerifiedStreamlet],
      inletPortPath: String
  ): Either[BlueprintProblem, VerifiedInlet] =
    for {
      verifiedPortPath ← VerifiedPortPath(inletPortPath)
      verifiedInlet ← verifiedStreamlets
        .find(_.name == verifiedPortPath.streamletRef)
        .toRight(PortPathNotFound(inletPortPath))
        .flatMap { verifiedStreamlet ⇒
          // a port path can leave out the inlet
          val portNameFound = verifiedPortPath.portName.orElse {
            if (verifiedStreamlet.descriptor.inlets.size == 1) Some(verifiedStreamlet.descriptor.inlets.head.name)
            else None
          }
          verifiedStreamlet.inletRefs
            .find(inletRef => Some(inletRef.inletName) == portNameFound)
            .map { inletRef =>
              Left(PortAlreadyBoundToTopic(inletPortPath, inletRef.topic))
            }
            .getOrElse {
              if (verifiedPortPath.portName.isEmpty && verifiedStreamlet.descriptor.inlets.size > 1) {
                val suggestions = verifiedStreamlet.descriptor.inlets.map(inlet ⇒ verifiedPortPath.copy(portName = Some(inlet.name)))
                Left(PortPathNotFound(inletPortPath, suggestions))
              } else {
                val portPath = if (verifiedPortPath.portName.isEmpty && verifiedStreamlet.descriptor.inlets.size == 1) {
                  verifiedPortPath.copy(portName = Some(verifiedStreamlet.descriptor.inlets.head.name))
                } else verifiedPortPath

                verifiedStreamlet.descriptor.inlets
                  .find(inlet ⇒ Some(inlet.name) == portPath.portName)
                  .map(inletDescriptor ⇒ VerifiedInlet(verifiedStreamlet, inletDescriptor.name, inletDescriptor.schema))
                  .toRight(PortPathNotFound(inletPortPath))
              }
            }
        }
    } yield verifiedInlet
}

final case class VerifiedInlet(streamlet: VerifiedStreamlet, portName: String, schemaDescriptor: SchemaDescriptor) extends VerifiedPort {
  def portPath = VerifiedPortPath(streamlet.name, Some(portName))
}

final case class VerifiedOutlet(streamlet: VerifiedStreamlet, portName: String, schemaDescriptor: SchemaDescriptor) extends VerifiedPort {
  def matches(outletDescriptor: OutletDescriptor) =
    outletDescriptor.name == portName &&
      outletDescriptor.schema.fingerprint == schemaDescriptor.fingerprint
  def portPath = VerifiedPortPath(streamlet.name, Some(portName))
}
